DataRepo/models.py

```
@@ -93,11 +93,64 @@ class Protocol(models.Model):
     category = models.CharField(
         max_length=256,
         choices=CATEGORY_CHOICES,
-        default=MSRUN_PROTOCOL,
         help_text="Classification of the protocol, "
         "e.g. an animal treatment or MSRun procedure.",
     )
 
+    @classmethod
+    def Retrieve_protocol(cls, **kwargs):
+        """
+        retrieve or insert a protocol, based on input.
+        protocol_input can either be a name or an integer (protocol_id)
+        """
+        protocol_input = kwargs.get("protocol_input")
+        category = kwargs.get("category")
+
+        print(f"Finding or inserting {category} protocol for '{protocol_input}'...")
+
+        action = "Found"
+
+        if cls.Is_integer(protocol_input):
+
+            try:
+                protocol = Protocol.objects.get(id=protocol_input)
+            except Protocol.DoesNotExist as e:
+                print("Protocol does not exist.")
+                raise e
+        else:
+            try:
+                protocol, created = Protocol.objects.get_or_create(
+                    name=protocol_input,
+                    category=category,
+                )
+
+                if created:
+                    action = "Created"
+                    # add the provisional description
+                    description = kwargs.get("provisional_description")
+                    if description is not None:
+                        protocol.description = description
+                        protocol.full_clean()
+                        protocol.save()
+
+            except Exception as e:
+                print(f"Failed to get or create {category} protocol {protocol_input}")
+                raise e
+
+        print(
+            f"{action} {category} protocol {protocol.id} '{protocol.name}' '{protocol.description}'"
+        )
+        return protocol
+
+    # may want to move this to a site-utility class, sometime
+    @classmethod
+    def Is_integer(cls, data):
+        try:
+            int(data)
```

NB. I'm not sure if casting to an integer is the best strategy.  I'm not certain, but it might silently indicate that `1.5` is an integer.  I would advocate using `isinstance(i, int)` instead. And if we use that, we probably wouldn't need a separate method.  Or, @lparsons may know a better method.

---

Happy to substitute a better solution, if proposed. Not really happy that with that `is_integer` method sort of floating around, for lack of a better place to put it or a built-in python equivalent that I could not find.

---

I believe you have a string here, so you need to cast to an `int`. However, I wouldn't both with the extra method, it doesn't really do anything. I'd just put the cast where you want it and catch the `ValueError` exception.

---

It was really just a means to an ends to get the logic to behave as I envisioned (`get` vs `get_or_create` dependent on input).   I sort of prefer it the way it reads (as-is), but if people want it gone, I can convert it.
https://stackoverflow.com/questions/1265665/how-can-i-check-if-a-string-represents-an-int-without-using-try-except

---

I would keep the try/except as that is pretty common in Python, in my experience. You can either do the cast and catch the exception or just perform the filter on id with the whatever the string is, and you can catch the exception then (which might even be cleaner).
```python
try:
    p = Protocol.objects.filter(id=protocol_input)
except ValueError:
    (p, created) = Protocol.objects.get_or_create(name=protocol_input)
```

---

Ah, right.  Strong typing.  I forgot.  So would we expect a value error on `int("1.5")` then, or does it silently convert it to `1`?

---

```python
$ python manage.py shell
Python 3.6.8 |Anaconda, Inc.| (default, Dec 30 2018, 01:22:34) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: int("1")
Out[1]: 1

In [2]: int("a")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-d9136db7b558> in <module>
----> 1 int("a")

ValueError: invalid literal for int() with base 10: 'a'

In [3]: int("1.5")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-aece86d34916> in <module>
----> 1 int("1.5")

ValueError: invalid literal for int() with base 10: '1.5'
```

---

re-coded to try a bunch of `try` `except` conditions